packages/backend/src/app/store-entry/store-entry.service.ts

```diff
@@ -1,36 +1,62 @@
 import { apId, ProjectId, PutStoreEntryRequest, StoreEntry } from '@activepieces/shared'
 import { databaseConnection } from '../database/database-connection'
 import { StoreEntryEntity } from './store-entry-entity'
+import { acquireLock, ApLock } from '../helper/lock'
 
 const storeEntryRepo = databaseConnection.getRepository<StoreEntry>(StoreEntryEntity)
 
 export const storeEntryService = {
     async upsert({ projectId, request }: { projectId: ProjectId, request: PutStoreEntryRequest }): Promise<StoreEntry | null> {
-        const previousEntry = await this.getOne({ projectId, key: request.key })
-        if (previousEntry !== null) {
-            await storeEntryRepo.update(previousEntry.id, request)
-            return this.getOne({ projectId, key: request.key })
-        }
-        else {
+        const keyLock = await getLock(request.key)
+
+        try {
+            const previousEntry = await this.getOne({ projectId, key: request.key })
+
+            if (previousEntry) {
+                await storeEntryRepo.update(previousEntry.id, request)
+
+                return await this.getOne({ projectId, key: request.key })
+            }
+
             const entryRequest: Omit<StoreEntry, 'created' | 'updated'> = {
                 id: apId(),
                 key: request.key,
                 value: request.value,
                 projectId,
             }
-            return storeEntryRepo.save(entryRequest)
+
+            return await storeEntryRepo.save(entryRequest)
```

Comment:
Use upsert, this suppose to be atomic

https://github.com/activepieces/activepieces/blob/09422ca992665ea9d2be4ad7c180a134e6dd8a56/packages/backend/src/app/ee/referrals/referral.service.ts#L28